docs/proposals/14-advanced-scheduling.md

```diff
@@ -0,0 +1,77 @@
+
+* if there is at least one un-tolerated taint with effect `NoSchedule` then Gardener will not schedule the shoot onto that seed
+* if there is no un-tolerated taint with effect `NoSchedule` but there is at least one un-tolerated taint with effect `PreferNoSchedule` then Gardener will try to not schedule the shoot onto the seed
+* if there is at least one un-tolerated taint with effect `NoExecute` then the shoot will be evicted from the seed (if it is already running on the seed), and will not be scheduled onto the seed (if it is not yet running on the seed).
+ 
+## Migrating Shoots Away from Unhealthy Seeds
+
+There are situations in which a seed can't function normally and should be either deleted or repaired "offline". This could be made easier by migrating shoots away from unhealthy seeds to healthy ones automatically or semi-automatically. This scenario is triggered by detecting an unhealthy seed. Once such a seed is detected, it should be drained of all its shoots, which means that they should be evicted from the seed and scheduled to other, healthy seeds.
+
+### Marking Seeds as "Unschedulable"
+
+The first step in migrating shoots away from an unhealthy seed is marking the seed as being "unschedulable" via a special taint, for example `seed.gardener.cloud/unschedulable`, with a `NoSchedule` effect. Initially, this could simply be done manually, or perhaps assisted by a special `gardenctl` command, for example `gardenctl taint <seed> `. In addition, the `gardenctl drain <seed>` command described below should add the taint to the specified seed prior to actually evicting all currently scheduled shoots.
+
+Marking a seed as "unschedulable" prevents any new shoots to be scheduled onto this seed, due to the `NoSchedule` effect of the taint.
+
+Later, detecting unhealthy seeds and marking them as "unschedulable" could also be done automatically. This could be achieved by extending the `gardenlet` seed controller to react on certain seed conditions by adding certain taints with `NoSchedule` effects. Such taints could include for example:
+
+* `seed.gardener.cloud/not-ready`: The seed is `NotReady`. 
+* `seed.gardener.cloud/over-capacity`: The seed's capacity for shoots has been exceeded, see [Ensuring Seeds Capacity for Shoots Is Not Exceeded](13-automated-seed-management.md#ensuring-seeds-capacity-for-shoots-is-not-exceeded).
+
+### Evicting Shoots from Drained Seeds
+
+The next step in migrating shoots away from an unhealthy seed is actually draining the seed, which involves evicting all shoots from this seed. This could be achieved in one of the following ways:
+
+* By manually marking all seed shoots as orphaned, either by adding an `orphan` annotation or simply be deleting the `seedName` field in the shoot specification.
```

Comment:
I think, unsetting `seedName` should be enough. We probably, don't need a second/separate indicator for shoots that should be moved/rescheduled.

Reply:
Yes, I agree, I only described it like this it because I am not sure unsetting the `seedName` and then setting it to a different value would work as expected right now (trigger the migration). This should however be tested and fixed.

Reply:
Removed the mention of an `orphan` annotation.